pennylane/pulse/jax_utils.py

```diff
@@ -0,0 +1,142 @@
+
+has_jax = True
+try:
+    import jax
+    import jax.numpy as jnp
+    from jax.experimental import sparse
+    from jax.tree_util import register_pytree_node_class
+except ImportError:
+    has_jax = False
+
+
+@register_pytree_node_class
+@dataclass
+class JaxParametrizedOperator:
+    """Jax pytree containing a parametrized operator."""
+
+    O_fixed: Optional[sparse.CSR]
+    O_parametrized: Tuple[sparse.CSR, ...]
+    coeff_parametrized: Tuple[
+        Callable[[float, float], complex], ...
+    ] = field()  # pytree_node=False)
+
+    def __call__(self, pars, t):
+        if self.O_fixed is not None:
+            ops = (self.O_fixed,)
+            coeffs = (1,)
+        else:
+            ops = ()
+            coeffs = ()
+        coeffs = coeffs + tuple(f(p, t) for f, p in zip(self.coeff_parametrized, pars))
+        return JaxLazyDot(coeffs, ops + self.O_parametrized)
+
+    @staticmethod
+    def from_parametrized_hamiltonian(
+        H: ParametrizedHamiltonian, dense: bool = False, wire_order=None
+    ):
+        """Construct a ``JaxParametrizedOperator`` from a ``ParametrizedHamiltonian``
+
+        Args:
+            H (ParametrizedHamiltonian): parametrized Hamiltonian to convert
+            dense (bool, optional): Decide wether a dense/sparse matrix is used. Defaults to False.
+            wire_order (list, optional): Wire order of the returned ``JaxParametrizedOperator``.
+                Defaults to None.
+
+        Returns:
+            JaxParametrizedOperator: jax-compatible parametrized operator
+        """
+        make_array = jnp.array if dense else sparse.CSR.fromdense
+
+        if len(H.ops_fixed) > 0:
+            fixed_mat = make_array(qml.matrix(H.H_fixed(), wire_order=wire_order))
+        else:
+            fixed_mat = None
+
+        parametrized_mat = tuple(
+            make_array(qml.matrix(op, wire_order=wire_order)) for op in H.ops_parametrized
+        )
+
+        return JaxParametrizedOperator(fixed_mat, parametrized_mat, tuple(H.coeffs_parametrized))
+
+    def tree_flatten(self):
+        """Function used by ``jax`` to flatten the JaxParametrizedOperator.
+
+        Returns:
+            tuple: tuple containing the matrices and the parametrized coefficients defining the class
+        """
+        matrices = (
+            self.O_fixed,
+            self.O_parametrized,
+        )
+        param_coeffs = self.coeff_parametrized
+        return (matrices, param_coeffs)
+
+    @classmethod
+    def tree_unflatten(cls, param_coeffs: tuple, matrices: tuple):
+        """Function used by ``jax`` to unflatten the JaxParametrizedOperator.
+
+        Args:
+            param_coeffs (tuple): tuple containing the parametrized coefficients of the class
+            matrices (tuple): tuple containing the matrices of the class
+
+        Returns:
+            JaxParametrizedOperator: a JaxParametrizedOperator instance
+        """
+        return cls(*matrices, param_coeffs)
```

Comment:
Note for the future (which I think we should do it anyway). By using a pytree library you can get rid of those two functions.

For example, if you have flax installed you could just define

```python
@flax.struct.dataclass
class JaxParametrizedOperator:
     """Jax pytree containing a parametrized operator."""

     O_fixed: Optional[sparse.CSR]
     O_parametrized: Tuple[sparse.CSR, ...]
     coeff_parametrized: Tuple[
         Callable[[float, float], complex], ...
     ] = flax.struct.field(pytree_node=False)
    ...
```

And the flattening/unflattening would be properly handled automatically (so no bugs for you!).

if pytrees are the future for Pennylane, I think it will eventually make sense to use a library for that. 
Options are: 
 - you bundle the logic to automatically define those methods yourself, as we do in [netket](https://github.com/netket/netket/blob/master/netket/utils/struct/dataclass.py). You could just take [this folder](https://github.com/netket/netket/tree/master/netket/utils/struct) and verndor it with minimal changes. NetKet also has the tests for it so it would be 0 effort.
 - rely on a library such as
   -  flax, though it's fairly large and ML oriented so unsure you'd want to depend on it.
   - [`treeo`](https://github.com/cgarciae/treeo), smaller lib works but it's maintained by only a guy (who I think works for google now)

For now I think it's fine to define those methods by hand, but in the future a choice should be made and a decorator should be used.

cc @josh146

Reply:
Also cc @trbromley. I wasn't aware of treeo, it looks really neat